packages/logger/index.js

```diff
@@ -0,0 +1,123 @@
+const logger = require('../dd-trace/src/log')
+// path to require tracer logger
+
+const https = require('https')
+const coalesce = require('koalas')
+
+class V2LogWriter {
+  // Note: these attribute names match the corresponding entry in the JSON payload.
+  constructor ({ ddsource, hostname, service, site, apiKey, interval, timeout }) {
+    this.ddsource = ddsource
+    this.hostname = hostname
+    this.service = service
+    this.interval = interval
+    this.timeout = timeout
+    this.buffer = []
+    this.buffer_limit = 1000
+    this.apiKey = apiKey
+    this.endpoint = '/api/v2/logs'
+    this.site = coalesce(site, 'datadoghq.com')
+    this.intake = `http-intake.logs.${this.site}`
+    this.headers = {
+      'DD-API-KEY': this.apiKey,
+      'Content-Type': 'application/json'
+    }
+    this.timer = setInterval(() => {
+      this.flush()
+    }, this.interval)
+
+    logger.debug(`"started log writer to ${this.url}"`)
+  }
+
+  start () {
+    this.timer()
+  }
+
+  tagString (tags) {
+    let tagStr = ''
+    for (const key in tags) {
+      tagStr += key + ':' + tags[key] + ','
+    }
+    return tagStr
+  }
+
+  log (log, tags, span) {
+    const logTags = this.tagString(tags)
+    const toLog = {
+      'timestamp': Date.now(),
+      'message': log.message,
+      'hostname': coalesce(log.hostname, this.hostname),
+      'ddsource': coalesce(log.ddsource, this.ddsource),
+      'service': coalesce(log.service, this.service),
+      'status': log.level,
+      'ddtags': logTags,
+      'dd.trace_id': span.trace_id + '',
+      'dd.span_id': span.span_id + ''
+    }
+    for (const key in log) {
+      if (!toLog[key]) {
+        toLog[key] = log[key]
+      }
+    }
+    return toLog
+  }
+
+  enqueue (log) {
+    if (this.buffer.length >= this.buffer_limit) {
+      logger.warn(`"log buffer full (limit is ${this.buffer_limit}), dropping log"`)
+      this.buffer = []
```

Comment:
This seems like it might not be the best approach to me.

Basically, once the buffer reaches the 1,000 entry limit, all of the 1,000 logs are thrown away and then only the new one survives. Instead, I think we would either want to throw away the oldest log (call `this.buffer.shift()`), or simply don't enqueue the newest log (simply call `return`).

If, for example, a service adds 101 logs per second, then we would hit 1,010 logs in 10 seconds, and the logs would be erased and maybe only the last 10 are sent, so about 99% are discarded. However, if we can hold onto 1,000 in some manner, and get rid of the other 10, then more like only 1% are discarded.

If the Python implementation did this then we might want to alert them to this scenario.

Reply:
In this case it should just return, so the new log would not be added to the queue, should there be an additional step in this case? Should the queue be flushed if the queue has become too big?

Reply:
I do like your idea of performing a flush if we've gotten too big. I suppose the queueing functionality isn't so much to prevent the backend from receiving too many logs, but is instead used to simply support batching of logs and reducing the overall send rate.

In that case, this might do it:

```js
  enqueue (log) {
    if (this.buffer.length >= this.buffer_limit) {
      this.flush() // makes a copy of this.buffer array and clears old array
    }
    this.buffer.push(log) // this is now cleared
  }
```

Then we wouldn't even need to log an error as there shouldn't be any dropped logs.

The only risk I can think of is that if user code generates a million logs then dd-trace will suddenly create 1,000 HTTP requests, which would slow down their app. That's probably a fine risk though since the user probably has other things to worry about at that point.